tests/AddressBookControllerIntegrationTest.java

```diff
@@ -0,0 +1,291 @@
+   *
+   */
+  @Test
+  public void removePersonToAddressBookSpy(){
+    controllerTest = new AddressBookController(addressBook_spy);
+    //Add a test person first
+    controllerTest.add(person_mock);
+    controllerTest.remove(0);
+    verify(addressBook_spy).remove(isA(int.class));
+    //Assert exception because arraylist but must be empty
+    assertThrows(IndexOutOfBoundsException.class,() ->addressBook_spy.get(0));
+  }
+
+
+
+  /**
+   * This method stubs a addressbook, to return a set value
+   * "person mock" to verify the method is called.
+   */
+  @Test
+  public void getPersonFromAddressBookStub(){
+    //Stub method, to return person_mock when called
+    when(addressBook_mock.get(0)).thenReturn(person_mock);
+
+    controllerTest.get(0); //Call test controller method
+    //verify method was called
+    verify(addressBook_mock).get(0);
+    //Verify return value is the same
+    assertEquals(person_mock, controllerTest.get(0));
+  }
+
+  /**
+   * This method address book spy and Person mock to verify
+   * that the addressbook gets the person's first and last name as toString
+   */
+  @Test
+  public void getPersonFromAddressBookSpy(){
+    //Initialize with spy
+    controllerTest = new AddressBookController(addressBook_spy);
+    //Add person mock
+    controllerTest.add(person_mock);
+    //Mock person's return value
+    when(person_mock.toString()).thenReturn("Bread, Ginger");
+    controllerTest.get(0);
+    verify(addressBook_spy).get(0);//Verify method is ran
+    //Assert that when controller test runs get(), then mocked return is called
+    assertEquals("Bread, Ginger",controllerTest.get(0).toString());
+  }
+
+  @Test
+  /**
+   * This method uses a addressbook mock to check if clear method is called.
+   */
+  public void clearAddressMock(){
+    doNothing().when(addressBook_mock).clear();
+    controllerTest.clear();
+    //verify method was called
+    verify(addressBook_mock).clear();
+  }
+
+  @Test
+  /**
+   * This method uses a addressbook spy to check
+   * if clear method is ran successfully.
+   */
+  public void clearAddressSpy(){
+    controllerTest= new AddressBookController(addressBook_spy);
+    addressBook_spy.add(person_mock);
+    controllerTest.clear();
+    verify(addressBook_spy).clear();
+    assertEquals(0,addressBook_spy.getPersons().length);
+  }
+
+
+
+  /**
+   * This method makes a spy of AddressBook and will throw
+   * an error when trying to save with an invalid file
+   */
+  @Test
+  public void throwSQLExceptionWhenSavingAddressBook() {
+    AddressBook addressBookSpy = spy(AddressBook.class);
+    AddressBookController controllerspy = spy(new AddressBookController(addressBookSpy));
+    try {
+      doThrow(SQLException.class).when(controllerspy).save(isA(File.class));
+      //controllerspy.save(new File("test"));
+
+      assertThrows(SQLException.class, () -> controllerspy.save(new File("test")));
+    } catch (Exception e) {
+      e.printStackTrace();
+    }
+  }
+  @Rule
```

Comment:
How come this declaration is down here? Can it be moved to the top of the file?

Reply:
fixed